### ISSUE TITLE
depexts uses 'osx' but filters use 'darwin'

### ISSUE THREAD

**Issue by dsheets** [ocaml/opam#2391]
An `opam` `depexts` field uses the name `osx` for the Apple operating system called OS X.

An `opam` `os` filter users the name `darwin` for the Apple operating system called OS X.

I understand that `darwin` may be technically correct in some sense as a distinction between the OS and the distribution but it is extremely confusing as there is only a single `darwin` distribution of note, OS X. I think that, at the least, `osx` should be introduced as an alias to `darwin` for `os` filter expressions. As it stands, not even a warning is emitted if you use `osx` in a filter (and obviously it does not trigger on OS X as opam believes that OS is called Darwin).

**Comment by AltGr** [comment]
Note that, at the moment, opam has no internal understanding of the contents of the `depexts:` field, which is defined by the repository (and by opam-depext, obviously).

Otherwise, it seems that `"Darwin"` was chosen because that's what `uname -s` returns (see https://github.com/ocaml/opam/blob/a8e79a514a790d4bf23810f1322f3af99971016e/src/core/opamStd.ml#L573-L580 ; is that always correct ?). But that would be quite easy to change. Having an alias would be more difficult, since filters evaluate as strings and comparisons thereof.

**Comment by dbuenzli** [comment]
I rather think `osx` should be dropped. All configury thing out there uses `Darwin` since that's what `uname -s` *always* return.

**Comment by dsheets** [comment]
If no alias is introduced, use of the unused name should become a warning or error.

Considering the future in which we can filter on OS versions, I would much rather write restrictions in terms of OS X version numbers (10.11.2) than Darwin version numbers (15.2.0). Further, the only testing that will be done will be on OS X and many of the packages talking about `darwin` will only work on OS X due to integration with Apple software which is not distributed with the open source Darwin.

In short, I believe calling the operating system `darwin` in an end-user developer context is unnecessarily confusing and very often wrong (unless you're talking about compilation targets perhaps).

Whatever the outcome, these names should be aligned and the use of the other name should produce errors/warnings.

**Comment by AltGr** [comment]
Really, I must admit I don't even know what the distinction between the two really is. It's difficult to forbid from the opam side (filters are, after all, just expressions with variables, comparison and logical operators) ; but a lint warning checking for references to `osx`, or direct comparisons thereof with the `os`variable, would'nt be difficult to add.

**Comment by dsheets** [comment]
OS X (now macOS) is the name of the operating system and Darwin is the name of the kernel. Essentially no one uses or cares about Darwin directly and everyone uses the OS X distribution of Darwin. One important point is that Darwin and OS X have different version schemes with no one caring about Darwin versions and everyone caring about OS X versions.

I think aliasing the names, as annoying as that is, and introducing a lint warning for `darwin` is the right user interface change.

**Comment by dsheets** [comment]
Alternatively, `darwin` should be made to check against the kernel (/version) and `osx` should check against the OS (/version). Then, they aren't quite aliases and both retain their actual meaning.

**Comment by AltGr** [comment]
I can add a lint check, and maybe rewrites to the repository, but I don't see how, while keeping some consistency, the variable `os` could have two values at once so that both `os = "osx"` and `os = "darwin"` are true.

Regarding `depext`, if you look at the code, we already have a notion of architecture, OS and distribution -- and not yet distribution revision (or OS, for that matter -- but I guess no one would care about the Linux kernel version either). What gets a bit weird is that, on OSX:
- the OS is  internally ``Darwin`, which leads to a flag `"osx"`
- the distribution is either `homebrew` or `macports`, which makes sense if you consider it's the packaging system that matters to `depext`

Regarding revisions, this would require a change in the format of `depexts`, since at the moment they are only binary tags.

**Comment by AltGr** [comment]
Defining variables `is-osx`, `osx-version` and `darwin-version` would be feasible though. This could actually be done by calls to `uname` from the main configuration file.

**Comment by AltGr** [comment]
It's now `os = macos` everywhere.
